### PR TITLE
test/e2e_node/remote: support pure SSH mode

### PR DESCRIPTION
**Pull request by jlebon**
Right now, `run_remote.go` only supports GCE instances. But actually
running the tests is completely independent of GCE and could work just
as well on any SSH-accessible machine.

This patch adds a new `--mode` switch, which defaults to `gce` for
backwards compatibility, but can be set to `ssh`. In that mode, the GCE
API is not used at all, and we simply connect to the hosts given via
`--hosts`.

This is still better than `run_local.go` because the latter mixes build
environment with test environment, which doesn't fit well with
container-optimized operating systems.

This is part of an effort to setup the e2e node tests on Fedora CoreOS
(see coreos/fedora-coreos-tracker#990).

Patch best viewed with whitespace ignored.

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

See commit message above.

#### Which issue(s) this PR fixes:

N/A

#### Special notes for your reviewer:

More context available in https://github.com/coreos/fedora-coreos-tracker/issues/990.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

N/A

### CODE REVIEW

**Review by endocrimes — COMMENTED** [review]
/assign

**Comment by jlebon** [comment]
I signed it.

**Comment by endocrimes** [comment]
@jlebon I've been on PTO, but if you ping me after a rebase I'll do a proper pass and test it out :) - Excited for this, should speed up my workflow a bunch too 😅

**Comment by jlebon** [comment]
@endocrimes Rebased! :)

File: test/e2e_node/runner/remote/run_remote.go
Comment (endocrimes) [inline test/e2e_node/runner/remote/run_remote.go:241]: I'd consider potentially doing a bigger refactor here..., in the new flow with `mode == ssh` it becomes fairly easy to provide configuration that would break, or potentially be confusing, and might be worth extracting state prep to manage that.

Comment (jlebon) [reply test/e2e_node/runner/remote/run_remote.go:315]: OK, pushed a couple of prep commits that splits things out a bit more!

**Comment by jlebon** [comment]
Rebased and updated for feedback!

File: test/e2e_node/runner/remote/run_remote.go
Comment (endocrimes) [reply test/e2e_node/runner/remote/run_remote.go:315]: Seems like a good enough first step to me 👍 

I'll give this a try

**Comment by jlebon** [comment]
@endocrimes Have you had a chance to take this for a spin?

**Review by endocrimes — APPROVED** [review]
image


/lgtm
/assign @ehashman @dims 

good first pass and start of the refactor is good. We might wanna spend some time generally cleaning up the runner but that's out of scope of this PR.

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm
/hold
/cc @SergeyKanzhelev @ehashman 
I'm fine with this merging as-is prior to test freeze but would like to hear from SIG Node so I'm holding off on adding the milestone

**Comment by BenTheElder** [comment]
should probably come from https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/OWNERS, I don't think @amwat is an owner in this repo.

build and hack changes seem OK.

for _cluster_ e2es we are avoiding assuming node SSH access, but I don't see an alternative for node e2es 🙃

**Comment by endocrimes** [comment]
My preference would be landing this as part of 1.23 for test workflow enhancements (and making it a little easier to continue with test cleanup), but final call is @ehashman / @SergeyKanzhelev

**Comment by spiffxp** [comment]
/milestone v1.23
/lgtm cancel
I'm going to add this to the milestone since I suspect that's why this hasn't been noticed yet, so instead I'm going to cancel my LGTM, and leave `/lgtm` and `/hold cancel` for @ehashman / @SergeyKanzhelev

**Comment by jlebon** [comment]
Rebased for conflicts!

File: hack/make-rules/test-e2e-node.sh
Comment (ehashman) [inline hack/make-rules/test-e2e-node.sh:206]: Follow-up: it looks like we may want to add the ability to specify a kubelet config file as well, I think that landed since this PR was written.

File: test/e2e_node/runner/remote/run_remote.go
Comment (ehashman) [inline test/e2e_node/runner/remote/run_remote.go:52]: I feel like everything else calls this "provider" but we can rename it later.

File: test/e2e_node/runner/remote/run_remote.go
Comment (ehashman) [inline test/e2e_node/runner/remote/run_remote.go:242]: Whitespace insensitive diff from new L240 down against old L321-389:

```diff
diff --git a/tmp/baz1 b/tmp/baz2
index f065ba9..dd10af1 100644
--- a/tmp/baz1
+++ b/tmp/baz2
@@ -1,6 +1,7 @@
                if *instanceNamePrefix == "" {
                        *instanceNamePrefix = "tmp-node-e2e-" + uuid.New().String()[:8]
                }
+       }
 
        // Setup coloring
        stat, _ := os.Stdout.Stat()
@@ -17,6 +18,7 @@
 
        results := make(chan *TestResult)
        running := 0
+       if gceImages != nil {
                for shortName := range gceImages.images {
                        imageConfig := gceImages.images[shortName]
                        fmt.Printf("Initializing e2e tests using image %s/%s/%s.\n", shortName, imageConfig.project, imageConfig.image)
@@ -25,6 +27,7 @@
                                results <- testImage(image, junitFileName)
                        }(&imageConfig, shortName)
                }
+       }
        if *hosts != "" {
                for _, host := range strings.Split(*hosts, ",") {
                        fmt.Printf("Initializing e2e tests using host %s.\n", host)
@@ -37,7 +40,6 @@
 
 
        // Wait for all tests to complete and emit the results
-       return gceImages, nil
        errCount := 0
        exitOk := true
        for i := 0; i < running; i++ {
```

**Review by ehashman — COMMENTED** [review]
/hold cancel
/lgtm

**Pull request merged**
